### PR TITLE
Specify that copy-out behavior still occurs after return/exit statements

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

**Comment by jafingerhut** [comment]
Note: I do NOT consider this to be a minor clarifying point in the spec.  I hope that this is NOT merged unless multiple p4c compiler developers think carefully on the issue, the questions raised in issue https://github.com/p4lang/p4-spec/issues/822, and how these proposed changes in the spec might affect various implementations.  In particular, whether an `exit` statement invoked from a control/action 3 or 4 levels deep in the control 'call stack' might lead to 'exit' being more expensive than currently believed.

**Comment by jnfoster** [comment]
Here's a question this issue made me consider for the first time: are an `exit` and a `return` in a top-level control semantically equivalent? 

An obvious and sensible interpretation is that they are.

Another possible interpretation would be that `exit` actually terminates the "thread" for the current packet. I don't think this is what was intended however.

If we decide that the cost of implementing the "obvious" semantics is too high, we could also imagine enriching the semantics to consider two different types of returns:
- Normal return (with copy-in / copy-out semantics)
- An abnormal return (with different semantics) due to an `exit` or `return`.

**Comment by jafingerhut** [comment]
When you say 'terminates the "thread" for the current packet' as a possible interpretation of `exit`, do you mean that the packet would be dropped?  If so, I am pretty sure that wasn't the intent for those who introduced `exit`.  I think the main motivation for adding both `return` and `exit` was having a way early in a long control to stop making any more changes to the state associated with the packet, without having to require the P4 developer to do something like more and more deeply nested if/then/else statements, and/or adding a "no_more_side_effects" Boolean variable that qualifies all future table apply statements.

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -4681,7 +4681,10 @@ The `return` statement immediately terminates the execution of the
 are not allowed within parsers.  `return` statements followed by an
 expression are only allowed within functions that return values; in
 this case the type of the expression must match the return type of the
-function.
+function.  Any copy-out behavior due to direction `out` or `inout`
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk]: "Enclosing" is rather a lexical concept, but you want to refer to the dynamically executing action or function here. I would rather say that before transferring control to the caller the out and inout parameter values are copied to the corresponding arguments.

**Comment by ChrisDodd** [comment]
I think its is desirable to preserve the property that call-by-reference and call-by-copy-in-out are equivalent if the argument is known not to alias with anything else accessed (directly or indirectly) in the callee.  So that if the compiler knows this (the argument is a value in a scope inaccessable to the the callee, or we analyze the callee to know this), it is free to elide copies and use references if that would be more efficient.

This would imply/require that copy-out does happen as part of exit/return, copying out the value of the local as of the exit/return statement.

File: p4-16/spec/P4-16-spec.mdk (hunk 1 above)
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: The first sentence of the paragraph also uses words that look like lexically enclosing thing, too: "The `return` statement immediately terminates the execution of the `action`, `function` or `control` containing it."

Do you want to change that sentence, too?  It seems to me that the lexically enclosing action, function, or control is exactly what `return` is meant to affect, isn't it?

Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk]: Ok, you can leave the text as it is.

**Pull request merged**